Sources/PackageLoading/ToolsVersionParser.swift

```diff
@@ -607,39 +605,47 @@ extension ManifestLoader {
+            // Otherwise, return the regular manifest.
+            return regularManifest
+        }
 
-            // SwiftPM 4 introduced tools-version designations; earlier packages default to tools version 3.1.0.
-            // See https://swift.org/blog/swift-package-manager-manifest-api-redesign.
-            let versionSpecificManifestToolsVersion: ToolsVersion
-            if versionSpecificCandidate < .v4 {
-                versionSpecificManifestToolsVersion = .v3
-            }
-            else {
-                versionSpecificManifestToolsVersion = try ToolsVersionParser.parse(manifestPath: versionSpecificManifest, fileSystem: fileSystem)
-            }
+        let versionSpecificManifest = packagePath.appending(component: versionSpecificManifests[versionSpecificCandidate]!)
 
-            // Try to get the tools version of the regular manifest.  At the comment marker is missing, we default to
-            // tools version 3.1.0 (as documented).
-            let regularManifestToolsVersion: ToolsVersion
-            do {
-                regularManifestToolsVersion = try ToolsVersionParser.parse(manifestPath: regularManifest, fileSystem: fileSystem)
-            }
-            catch let error as UnsupportedToolsVersion where error.packageToolsVersion == .v3 {
-              regularManifestToolsVersion = .v3
-            }
+        // SwiftPM 4 introduced tools-version designations; earlier packages default to tools version 3.1.0.
+        // See https://swift.org/blog/swift-package-manager-manifest-api-redesign.
+        let versionSpecificManifestToolsVersion: ToolsVersion
+        if versionSpecificCandidate < .v4 {
+            versionSpecificManifestToolsVersion = .v3
+        }
+        else {
+            versionSpecificManifestToolsVersion = try ToolsVersionParser.parse(manifestPath: versionSpecificManifest, fileSystem: fileSystem)
+        }
 
-            // Compare the tools version of this manifest with the regular
-            // manifest and use the version-specific manifest if it has
-            // a greater tools version.
-            if versionSpecificManifestToolsVersion > regularManifestToolsVersion {
+        // Compare the tools version of this manifest with the regular
+        // manifest and use the version-specific manifest if it has
+        // a greater tools version.
+        if versionSpecificManifestToolsVersion > regularManifestToolsVersion {
+            return versionSpecificManifest
+        } else {
+            // If there's no primary candidate, validate the regular manifest.
+            if regularManifestToolsVersion.validateToolsVersion(currentToolsVersion) {
+                return regularManifest
+            } else {
+                // If that's incompatible, use the closest version-specific manifest we got.
                 return versionSpecificManifest
```

Comment:
This is the key change, previously we would have always returned the regular manifest here, even if it was incompatible.

Reply:
Makes sense to me.  And we wound't even be in this part of the code unless we knew we had a `versionSpecificManifest`, right, since the `guard` above would have returned the regular manifest previously?

This logic makes sense to me.